File: frontend/app/types/property.ts
export interface PropertyInput {
  address: string
  suburb: string
  state: string
  postcode: string
  property_type: string
  bedrooms: number
  bathrooms: number
  car_spaces: number
  purchase_price: number
  estimated_rent_per_week: number
  loan_amount: number | null
  interest_rate: number
  loan_term_years: number
  is_new_build: boolean
  year_built: number | null
}

export interface RentalYieldResult {
  gross_yield_pct: number
  net_yield_pct: number
  estimated_vacancy_rate_pct: number
  annual_rental_income: number
  market_rent_assessment: string
  commentary: string
}

export interface CashflowResult {
  weekly_rental_income: number
  weekly_mortgage_payment: number
  weekly_expenses: number
  weekly_net_cashflow: number
  annual_net_cashflow: number
  is_positive_cashflow: boolean
  break_even_rent: number
  commentary: string
}

export interface ROIResult {
  estimated_capital_growth_pct_pa: number
  total_return_pct_pa: number
  equity_in_5_years: number
  projected_value_5_years: number
  payback_period_years: number | null
  commentary: string
}

export interface LocationRiskResult {
  suburb_score: number
  flood_risk: string
  crime_risk: string
  vacancy_risk: string
  infrastructure_score: number
  demand_supply_balance: string
  overall_risk_level: string
  key_drivers: string[]
  commentary: string
}

export interface TaxDepreciationResult {
  marginal_tax_rate_pct: number
  annual_depreciation_total: number
  division_40_depreciation: number
  division_43_depreciation: number
  annual_tax_deductible_loss: number
  estimated_annual_tax_benefit: number
  after_tax_weekly_cashflow: number
  is_negatively_geared: boolean
  commentary: string
}

export interface InvestmentPotentialResult {
  verdict: string
  confidence: string
  overall_score: number
  key_strengths: string[]
  key_risks: string[]
  recommendation: string
}

export interface NegotiationResult {
  asking_price: number
  recommended_max_offer: number
  suggested_opening_offer: number
  walk_away_price: number
  estimated_savings_potential: number
  market_position: string
  negotiation_levers: string[]
  confidence: string
  strategy: string
}

export interface PropertyReport {
  property: PropertyInput
  // Each result is optional — users can pick which analyses to run.
  rental_yield?: RentalYieldResult | null
  cashflow?: CashflowResult | null
  roi?: ROIResult | null
  location_risk?: LocationRiskResult | null
  tax_depreciation?: TaxDepreciationResult | null
  investment_potential?: InvestmentPotentialResult | null
  negotiation?: NegotiationResult | null
  selected_analyses?: string[] | null
  generated_at: string
  tokens_used: number
}

// Canonical analysis identifiers — must match backend ANALYSIS_IDS.
export type AnalysisId =
  | 'rental_yield'
  | 'cashflow'
  | 'roi'
  | 'location_risk'
  | 'tax_depreciation'
  | 'investment_potential'
  | 'negotiation'

export const ALL_ANALYSES: AnalysisId[] = [
  'rental_yield',
  'cashflow',
  'roi',
  'location_risk',
  'tax_depreciation',
  'investment_potential',
  'negotiation',
]

export type StreamEvent =
  | { event: 'rental_yield'; data: RentalYieldResult }
  | { event: 'cashflow'; data: CashflowResult }
  | { event: 'roi'; data: ROIResult }
  | { event: 'location_risk'; data: LocationRiskResult }
  | { event: 'tax_depreciation'; data: TaxDepreciationResult }
  | { event: 'investment_potential'; data: InvestmentPotentialResult }
  | { event: 'negotiation'; data: NegotiationResult }
  | { event: 'complete'; data: PropertyReport }
  | { event: 'saved'; data: { report_id: string } }
  | { event: 'error'; data: { message: string } }

File: backend/app/schemas/property.py
from __future__ import annotations
from datetime import datetime, UTC
from typing import Optional
from pydantic import BaseModel, Field


class PropertyInput(BaseModel):
    address: str
    suburb: str
    state: str
    postcode: str
    property_type: str = Field(..., description="house | unit | townhouse | land")
    bedrooms: int
    bathrooms: int
    car_spaces: int = 0
    purchase_price: float
    estimated_rent_per_week: float
    loan_amount: Optional[float] = None
    interest_rate: float = 6.5
    loan_term_years: int = 30
    is_new_build: bool = False
    year_built: Optional[int] = None


class RentalYieldResult(BaseModel):
    gross_yield_pct: float
    net_yield_pct: float
    estimated_vacancy_rate_pct: float
    annual_rental_income: float
    market_rent_assessment: str  # below_market | at_market | above_market
    commentary: str


class CashflowResult(BaseModel):
    weekly_rental_income: float
    weekly_mortgage_payment: float
    weekly_expenses: float
    weekly_net_cashflow: float
    annual_net_cashflow: float
    is_positive_cashflow: bool
    break_even_rent: float
    commentary: str


class ROIResult(BaseModel):
    estimated_capital_growth_pct_pa: float
    total_return_pct_pa: float
    equity_in_5_years: float
    projected_value_5_years: float
    payback_period_years: Optional[float]
    commentary: str


class LocationRiskResult(BaseModel):
    suburb_score: int = Field(..., ge=1, le=10)
    flood_risk: str  # low | medium | high
    crime_risk: str  # low | medium | high
    vacancy_risk: str  # low | medium | high
    infrastructure_score: int = Field(..., ge=1, le=10)
    demand_supply_balance: str  # undersupply | balanced | oversupply
    overall_risk_level: str  # low | medium | high
    key_drivers: list[str]
    commentary: str


class InvestmentPotentialResult(BaseModel):
    verdict: str  # BUY | HOLD | AVOID
    confidence: str  # high | medium | low
    overall_score: int = Field(..., ge=1, le=10)
    key_strengths: list[str]
    key_risks: list[str]
    recommendation: str


class PropertyReport(BaseModel):
    property: PropertyInput
    rental_yield: RentalYieldResult
    cashflow: CashflowResult
    roi: ROIResult
    location_risk: LocationRiskResult
    investment_potential: InvestmentPotentialResult
    generated_at: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())
    tokens_used: int = 0


class AnalyzeRequest(BaseModel):
    property: PropertyInput


class AnalyzeResponse(BaseModel):
    report: PropertyReport
